adobe/copy_on_write.hpp

```diff
@@ -55,93 +36,111 @@ copy when the value is modified and there is more than one reference to the
 
     /*!
-        Initializes the instance with a custom allocator
+    \brief Constructs a new copy_on_write object with a single argument.
     */
-    explicit copy_on_write(const allocator_type& a) : object_m(0) {
-        implementation_allocator_type other_allocator(a);
-
-        object_m = allocate(other_allocator);
-    }
+    template <class U>
+    copy_on_write(U&& x, disable_copy<U> = nullptr) : _self(new model(std::forward<U>(x))) {}
 
     /*!
-    \brief Constructs a new copy_on_write object with a value \c x.
-
-    \param x A default value to assign to this object
+    \brief Constructs a new copy_on_write object with two or more arguments.
     */
-    copy_on_write(T x) : object_m(allocate_move(0, std::move(x))) {}
+    template <class U, class V, class... Args>
+    copy_on_write(U&& x, V&& y, Args&&... args)
+        : _self(new model(std::forward<U>(x), std::forward<V>(y), std::forward<Args>(args)...)) {}
 
     /*!
     Copy construction is a non-throwing operation and simply increments the
     reference count on the stored object.
     */
-    copy_on_write(const copy_on_write& x) : object_m(x.object_m) {
-        if (object_m)
-            object_m->header_m.get().count_m.increment();
-    }
+    copy_on_write(const copy_on_write& x) noexcept : _self(x._self) {
+        assert(_self && "FATAL (sparent) : using a moved copy_on_write object");
 
-    copy_on_write(copy_on_write&& x) : object_m(x.object_m) { x.object_m = 0; }
+        ++_self->_count;
+    }
+    copy_on_write(copy_on_write&& x) noexcept : _self(x._self) {
+        assert(_self && "WARNING (sparent) : using a moved copy_on_write object");
+        x._self = nullptr;
+    }
 
-    ~copy_on_write() { release(object_m); }
+    ~copy_on_write() { destroy(); }
 
     /*!
         As with copy construction, assignment is a non-throwing operation which
         releases the old value and increments the reference count of the item
         being assigned to.
     */
-    copy_on_write& operator=(copy_on_write x) {
-        swap(*this, x);
-        return *this;
+    auto operator=(const copy_on_write& x) noexcept -> copy_on_write& {
+        return *this = copy_on_write(x);
     }
+    auto operator=(copy_on_write&& x) noexcept -> copy_on_write& {
+        assert(x._self && "WARNING (sparent) : using a moved copy_on_write object");
+        assert((&x != this) && "FATAL (sparent) : invalid self-move.");
 
+        destroy();
+        _self = x._self;
+        x._self = nullptr;
+        return *this;
+    }
 
-    copy_on_write& operator=(T x) {
-        if (!object_m)
-            object_m = allocate_move(0, std::move(x));
-        else if (object_m->header_m.get().count_m.is_one())
-            object_m->value_m = std::move(x);
-        else
-            reset(allocate_move(object_m, std::move(x)));
+    template <class U>
+    auto operator=(U&& x) -> disable_copy_assign<U> {
+        if (_self && unique()) {
+            _self->_value = std::forward<U>(x);
```

Comment:
Remind me again why it isn't possible that another thread can come in and bump the ref count of the underlying model between unique and the assignment? Is it because the two threads would be expected to have their own cow instances, and the ref count would then be at least two?